proposals/p6333.md

```diff
@@ -0,0 +1,586 @@
+`cc_library`.
+
+### Support a full-fledged build system
+
+The `build` command as proposed here is intended to be sufficient for quick
+testing and simple tools. However, it's not intended to be flexible with custom
+rules, plugins, and so on. These are features offered by systems such as CMake
+or Bazel.
+
+Instead, we could provide a full build system. Multiple other languages have
+gone in that direction:
+
+-   In Rust, `cargo` combines a
+    [build system](https://doc.rust-lang.org/cargo/commands/cargo-build.html)
+    and package manager.
+-   In Swift,
+    [SwiftPM](https://www.swift.org/documentation/server/guides/building.html)
+    provides a similar offering as to `cargo`.
+-   In Zig, there are
+    [multiple build system](https://ziglang.org/learn/build-system/) commands.
+
+Carbon's
+[project goal](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+is migration of existing C++ developers, particularly "This means integrating
+into the existing C++ ecosystem by supporting incremental migration from C++ to
+Carbon."
+
+The expectation is that C++ users will already be using a fully featured build
+system, such as CMake. Migration should be easier if users can retain their
+existing build system, particularly since a typical migration can be expected to
+mix both Carbon and C++ code.
+
+While Carbon could provide _both_ a separate compilation system _and_ a fully
+featured build system, a build system is a substantial undertaking and we expect
+C++ developers to already have one.
+
+### Don't support packaging directive to filename mappings
+
+Instead of making a mapping from packaging directives to filenames, we could
+generate a list specific to the `Core` package, and not expose that for other
+packages.
+
+We shouldn't manually maintain a mapping for the `Core` package; it should be
+automated. It's likely that whatever we do in this space, however we would
+support a mapping, would be of interest to small projects. It will probably be
+low cost for us to build support things other than `Core`, so we should just do
+that.
+
+### Distribute pre-compiled versions of Core files
+
+Instead of building object files for `Core` on demand, we could distribute them
+as part of Carbon. The upside of this is it would make builds a little faster;
+the downside is that we'd end up in more of a situation where supported target
+platforms were enumerated, or perhaps where special platforms could be built
+on-demand in a bespoke manner.
+
+We can probably add limited caching where it'd help, and support all platforms
+using similar logic that way with little performance penalty.
+
+### Require C++ files be passed to clang directly
+
+We could require C++ files to be passed to clang or carbon clang instead of
+carbon build, and similarly cc_library instead of carbon_library. Behind the
+scenes, clang will be equivalent to carbon clang, so Carbon will already be able
+to have control over how that compiles.
```

Comment:
```suggestion
We could require C++ files to be passed to `clang` or `carbon clang` instead of
`carbon build`, and similarly `cc_library` instead of `carbon_library`. Behind the
scenes, `clang` will be equivalent to `carbon clang`, so Carbon will already be able
to have control over how that compiles.
```